File: frontend/src/types/voting.ts
export type AssemblyType = "AGO" | "AGE";
export type AssemblyStatus = "DRAFT" | "OPEN" | "CLOSED";
export type VoteKind = "ASSEMBLEIA" | "ENQUETE";
export type VoteType = "SINGLE_CHOICE" | "MULTIPLE_CHOICE";
export type VoteStatus = "OPEN" | "CLOSED";

export interface AssemblyRead {
  id: string;
  title: string;
  type: AssemblyType;
  held_on: string;
  agenda?: string | null;
  status: AssemblyStatus;
  created_by_id: string;
  created_at: string;
  updated_at: string;
  closed_at?: string | null;
}

export interface AssemblyCreatePayload {
  title: string;
  type: AssemblyType;
  held_on: string;
  agenda?: string | null;
}

export interface AssemblyUpdatePayload {
  title?: string;
  type?: AssemblyType;
  held_on?: string;
  agenda?: string | null;
  /**
   * CLOSED is not reachable through PATCH: closing cascades into every open
   * vote and freezes each tally snapshot, so it only happens through
   * `POST /assemblies/{id}/close`. The API rejects it with 400.
   */
  status?: Exclude<AssemblyStatus, "CLOSED">;
}

export interface VoteOptionRead {
  id: string;
  vote_id: string;
  label: string;
  order_index: number;
}

export interface VoteOptionCreatePayload {
  label: string;
  order_index?: number;
}

export interface VoteRead {
  id: string;
  assembly_id?: string | null;
  kind: VoteKind;
  title: string;
  description?: string | null;
  vote_type: VoteType;
  status: VoteStatus;
  is_anonymous: boolean;
  opens_at: string;
  closes_at: string;
  created_by_id: string;
  created_at: string;
  updated_at: string;
  closed_at?: string | null;
  options: VoteOptionRead[];
}

export interface VoteCreatePayload {
  assembly_id?: string | null;
  kind: VoteKind;
  title: string;
  description?: string | null;
  vote_type: VoteType;
  is_anonymous?: boolean;
  opens_at?: string | null;
  closes_at: string;
  options: VoteOptionCreatePayload[];
}

export interface VoteListParams {
  kind?: VoteKind;
  status?: VoteStatus;
  assembly_id?: string;
}

export interface BallotRead {
  id: string;
  vote_id: string;
  lot_id?: string | null;
  lot_label?: string | null;
  voter_user_id?: string | null;
  voter_name?: string | null;
  is_retraction: boolean;
  selected_option_ids: string[];
  cast_at: string;
}

/**
 * A ballot the caller is entitled to see. `can_edit` is true only for the
 * person who actually cast it — only that person may change or retract it.
 */
export interface MyBallotRead extends BallotRead {
  can_edit: boolean;
}

export interface BallotCreatePayload {
  lot_id?: string | null;
  selected_option_ids: string[];
}

export interface BallotRetractPayload {
  lot_id?: string | null;
}

export interface TallyResultRead {
  option_id: string;
  label: string;
  count: number;
}

export interface TallyAttributionRead {
  lot_id?: string | null;
  lot_label?: string | null;
  voter_user_id?: string | null;
  voter_name?: string | null;
  selected_labels: string[];
  cast_at: string;
}

/**
 * Two-mode payload: while the vote is open the API sends only
 * `voters_count` (plus `total_lots` in an assembly) and omits `results`
 * and `attributions` entirely.
 */
export interface TallyRead {
  vote_id: string;
  kind: VoteKind;
  status: VoteStatus;
  is_anonymous: boolean;
  voters_count: number;
  total_lots?: number;
  results?: TallyResultRead[];
  attributions?: TallyAttributionRead[];
}

export interface EligibleLotRead {
  id: string;
  label: string;
}

export interface LotVoterEligibilityRead {
  id: string;
  lot_id: string;
  user_id: string;
  user_name?: string | null;
  added_by_id?: string | null;
  added_at: string;
}

export interface LotVoterEligibilityCreatePayload {
  user_id: string;
}

File: backend/app/schemas/voting.py
"""Pydantic schemas for assemblies, votes, ballots and tallies (APRAS-33)."""

from datetime import date, datetime
from uuid import UUID

from app.models.enums import (
    AssemblyStatus,
    AssemblyType,
    VoteKind,
    VoteStatus,
    VoteType,
)
from pydantic import BaseModel, ConfigDict, Field, model_validator


# ---------------------------------------------------------------------------
# Assembly
# ---------------------------------------------------------------------------


class AssemblyCreate(BaseModel):
    title: str = Field(..., min_length=1)
    type: AssemblyType
    held_on: date
    agenda: str | None = None


class AssemblyUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1)
    type: AssemblyType | None = None
    held_on: date | None = None
    agenda: str | None = None
    status: AssemblyStatus | None = None


class AssemblyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    type: AssemblyType
    held_on: date
    agenda: str | None = None
    status: AssemblyStatus
    created_by_id: UUID
    created_at: datetime
    updated_at: datetime
    closed_at: datetime | None = None


# ---------------------------------------------------------------------------
# Vote / VoteOption
# ---------------------------------------------------------------------------


class VoteOptionCreate(BaseModel):
    label: str = Field(..., min_length=1)
    order_index: int = 0


class VoteOptionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    vote_id: UUID
    label: str
    order_index: int


class VoteCreate(BaseModel):
    assembly_id: UUID | None = None
    kind: VoteKind
    title: str = Field(..., min_length=1)
    description: str | None = None
    vote_type: VoteType
    is_anonymous: bool = False
    opens_at: datetime | None = None
    closes_at: datetime
    options: list[VoteOptionCreate] = Field(..., min_length=2)

    @model_validator(mode="after")
    def check_assembly_binding(self) -> "VoteCreate":
        """An ASSEMBLEIA vote belongs to an assembly; an ENQUETE never does."""
        if self.kind == VoteKind.ASSEMBLEIA and self.assembly_id is None:
            raise ValueError("assembly_id is required for ASSEMBLEIA votes")
        if self.kind == VoteKind.ENQUETE and self.assembly_id is not None:
            raise ValueError("assembly_id must be null for ENQUETE votes")
        return self


class VoteUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1)
    description: str | None = None
    vote_type: VoteType | None = None
    is_anonymous: bool | None = None
    opens_at: datetime | None = None
    closes_at: datetime | None = None
    options: list[VoteOptionCreate] | None = Field(default=None, min_length=2)


class VoteRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    assembly_id: UUID | None = None
    kind: VoteKind
    title: str
    description: str | None = None
    vote_type: VoteType
    status: VoteStatus
    is_anonymous: bool
    opens_at: datetime
    closes_at: datetime
    created_by_id: UUID
    created_at: datetime
    updated_at: datetime
    closed_at: datetime | None = None
    options: list[VoteOptionRead] = []


# ---------------------------------------------------------------------------
# Lot voter eligibility
# ---------------------------------------------------------------------------


class EligibleLotRead(BaseModel):
    """A lot the caller may cast the assembly ballot for.

    Needed by the ballot UI: a RESIDENT cannot list lots, so the vote
    screen has no other way to offer the right Bloco/Lote to vote for.
    """

    id: UUID
    label: str


class LotVoterEligibilityCreate(BaseModel):
    user_id: UUID


class LotVoterEligibilityRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    lot_id: UUID
    user_id: UUID
    user_name: str | None = None
    added_by_id: UUID | None = None
    added_at: datetime


# ---------------------------------------------------------------------------
# Ballots
# ---------------------------------------------------------------------------


class BallotCreate(BaseModel):
    lot_id: UUID | None = None
    selected_option_ids: list[UUID] = Field(..., min_length=1)


class BallotRetract(BaseModel):
    """Body of ``POST /votes/{id}/ballots/retract``.

    Carries the lot whose active ballot is being withdrawn: required in
    ``ASSEMBLEIA`` (a user may hold ballots for several lots in the same
    vote), and must be null/absent in ``ENQUETE``.
    """

    lot_id: UUID | None = None


class BallotRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    vote_id: UUID
    lot_id: UUID | None = None
    lot_label: str | None = None
    voter_user_id: UUID | None = None
    voter_name: str | None = None
    is_retraction: bool = False
    selected_option_ids: list[UUID] = []
    cast_at: datetime


class MyBallotRead(BallotRead):
    """One ballot the caller is entitled to see (see spec Visibilidade, Regra 3).

    ``can_edit`` is true only for the person who cast that specific ballot —
    only the holder may change or retract it.
    """

    can_edit: bool = False


# ---------------------------------------------------------------------------
# Tally
# ---------------------------------------------------------------------------


class TallyResultRead(BaseModel):
    option_id: UUID
    label: str
    count: int


class TallyAttributionRead(BaseModel):
    lot_id: UUID | None = None
    lot_label: str | None = None
    voter_user_id: UUID | None = None
    voter_name: str | None = None
    selected_labels: list[str] = []
    cast_at: datetime


class TallyRead(BaseModel):
    """Two-mode tally payload.

    While the vote is open only ``voters_count`` (plus ``total_lots`` in an
    assembly) is filled; ``results`` and ``attributions`` stay ``None`` and
    are dropped from the JSON response by ``response_model_exclude_none``.
    """

    vote_id: UUID
    kind: VoteKind
    status: VoteStatus
    is_anonymous: bool
    voters_count: int
    total_lots: int | None = None
    results: list[TallyResultRead] | None = None
    attributions: list[TallyAttributionRead] | None = None
